### PR TITLE
wallet: add gettransaction

### PR DESCRIPTION
**Pull request by ErikEk**
This PR adds a method to the wallet that allows to get transaction data on any transaction given its ID. Useful for (https://github.com/lightningnetwork/lnd/pull/7654).

### CODE REVIEW

File: wallet/wallet.go [hunk 1]
```diff
@@ -2464,6 +2464,73 @@ func (w *Wallet) GetTransactions(startBlock, endBlock *BlockIdentifier,
...
+			blockHash = &txDetail.Block.Hash
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var res GetTransactionResult
```
Comment (guggero) [inline wallet/wallet.go:2515]: Have this be the only variable outside of the `walletdb.View` call and move the `switch` below inside the callback? That way it's more clear what the result will be if `txDetail` is nil.

**Review by guggero — COMMENTED** [review]
Thanks for the addition, should definitely be useful for different use cases in `lnd`.

File: wallet/wallet.go [hunk 2]
```diff
@@ -2464,6 +2464,73 @@ func (w *Wallet) GetTransactions(startBlock, endBlock *BlockIdentifier,
...
+// structure for a mined transaction.
+type GetTransactionResult struct {
+	MinedTransaction   Block
+	UnminedTransaction TransactionSummary
+}
+
+// GetTransaction returns data of any transaction given its id. A mined
+// transaction is returned in a Block structure which records properties about
+// the block.
+func (w *Wallet) GetTransaction(txHash *chainhash.Hash) (*GetTransactionResult,
```
[the remark this answers is not in the document]
Comment (ErikEk) [reply wallet/wallet.go:2478]: Yes, that makes sence.

File: wallet/wallet.go (hunk 1 above)
Comment (ErikEk) [reply wallet/wallet.go:2515]: Great. I'll do that and return an error if the transaction isn't found.

Comment (ErikEk) [reply wallet/wallet.go]: Done. Now I return an error if the transaction isn't found instead.

File: wallet/wallet.go [hunk 3]
```diff
@@ -2464,6 +2464,66 @@ func (w *Wallet) GetTransactions(startBlock, endBlock *BlockIdentifier,
...
+			res.MinedTransaction = Block{
+				Hash:         blockHash,
+				Height:       bestHeight,
+				Timestamp:    summary.Timestamp,
+				Transactions: []TransactionSummary{summary},
+			}
+		}
+
+		return nil
+	})
```
Comment (guggero) [inline wallet/wallet.go:2522]: I think we should do an explicit `return nil, err` if the err is not nil here.

File: wallet/wallet.go [hunk 4]
```diff
@@ -2464,6 +2464,66 @@ func (w *Wallet) GetTransactions(startBlock, endBlock *BlockIdentifier,
...
+		if txDetail == nil {
+			return fmt.Errorf("can not find the transaction")
+		}
+
+		// Otherwise, we create a summary of the transaction details.
+		summary := makeTxSummary(dbtx, w, txDetail)
+
+		var bestHeight int32
+		var blockHash *chainhash.Hash
+		// If it is a confirmed transaction we set the corresponding
```
Comment (guggero) [inline wallet/wallet.go:2500]: nit: add newline before comment.

File: wallet/wallet.go [hunk 5]
```diff
@@ -2464,6 +2464,66 @@ func (w *Wallet) GetTransactions(startBlock, endBlock *BlockIdentifier,
...
+
+		// If the transaction was not found we return an error.
+		if txDetail == nil {
+			return fmt.Errorf("can not find the transaction")
+		}
+
+		// Otherwise, we create a summary of the transaction details.
+		summary := makeTxSummary(dbtx, w, txDetail)
+
+		var bestHeight int32
```
Comment (guggero) [inline wallet/wallet.go:2498]: nit: could group multiple `var` statements in a single block.
for example:
```
var (
    bestHeight int32
    blockHash *chainhash.Hash
    summary = makeTxSummary(...)
)
```

File: wallet/wallet.go [hunk 6]
```diff
@@ -2464,6 +2464,66 @@ func (w *Wallet) GetTransactions(startBlock, endBlock *BlockIdentifier,
...
+// structure for a mined transaction.
+type GetTransactionResult struct {
+	MinedTransaction   Block
+	UnminedTransaction TransactionSummary
+}
+
+// GetTransaction returns data of any transaction given its id. A mined
+// transaction is returned in a Block structure which records properties about
+// the block.
+func (w *Wallet) GetTransaction(txHash chainhash.Hash) (*GetTransactionResult,
```
Comment (guggero) [inline wallet/wallet.go:2478]: Can we add a unit test for this please?

**Review by guggero — COMMENTED** [review]
Getting pretty close!
Main ask is a unit test.

File: wallet/wallet.go (hunk 6 above)
Comment (ErikEk) [reply wallet/wallet.go:2478]: Alright. Will add a test.

**Comment by yyforyongyu** [comment]
feel free to ping me when it's ready for review

**Comment by ErikEk** [comment]
@yyforyongyu Just pushed the changes. Please take a look.

File: wallet/wallet.go [hunk 7]
```diff
@@ -2464,6 +2464,69 @@ func (w *Wallet) GetTransactions(startBlock, endBlock *BlockIdentifier,
 	return &res, err
 }
 
+// GetTransactionResult returns a transaction in the UnminedTransaction
+// field if it is unmined, or in the MinedTransaction field as a Block
+// structure for a mined transaction.
+type GetTransactionResult struct {
```
Comment (yyforyongyu) [inline wallet/wallet.go:2470]: Just thinking about the use case of this method, it's easier if we just give a flat result back, something like,
```go
type GetTransactionResult struct {
	TransactionSummary

	// Height...
	Height int32

	// BlockHash...
	BlockHash chainhash.Hash
}
```
Callers can check the `Height` field to decide whether it's mined or not, or, if we wanna be explicit, we can add one more field, something like,

```go
type GetTransactionResult struct {
	TransactionSummary

	// Height...
	Height int32

	// BlockHash...
	BlockHash chainhash.Hash

	// Confirmations...
	Confirmations uint32
```

What do you think? This way we can also get rid of the field `Transactions: []TransactionSummary{summary}` which can be confusing.

File: wallet/wallet_test.go [hunk 8]
```diff
@@ -205,3 +206,43 @@ func TestLabelTransaction(t *testing.T) {
...
+	t.Parallel()
+
+	w, cleanup := testWallet(t)
+	defer cleanup()
+
+	// Create a new tx record.
+	rec, err := wtxmgr.NewTxRecord(
+		TstSerializedTx, time.Now(),
+	)
+	if err != nil {
```
Comment (yyforyongyu) [inline wallet/wallet_test.go:222]: We are pushing the usage of `require` atm, so we wanna do `require.NoError` here and below.

File: wallet/wallet_test.go [hunk 9]
```diff
@@ -205,3 +206,43 @@ func TestLabelTransaction(t *testing.T) {
 		})
 	}
 }
+
+// TestGetTransaction writes a tx to disk and then compares
+// the hash to the response from the GetTransaction function.
+func TestGetTransaction(t *testing.T) {
```
Comment (yyforyongyu) [inline wallet/wallet_test.go:212]: The test could be more comprehensive, I think we should at least test three cases,
- when it's not found
- when it's not mined
- when it's confirmed

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Thanks for the PR! Left some comments about tests and the returned format.

File: wallet/wallet_test.go (hunk 9 above)
Comment (ErikEk) [reply wallet/wallet_test.go:212]: Makes sense. I will add them.

File: wallet/wallet.go (hunk 7 above)
Comment (ErikEk) [reply wallet/wallet.go:2470]: The rationale was to reuse the same importing technique as ListTransactionDetails uses in lnd (https://github.com/lightningnetwork/lnd/blob/91d910872d9a5d1490026a3a9f92a7afaa9af069/lnwallet/btcwallet/btcwallet.go#L1450).

Comment (ErikEk) [reply wallet/wallet.go:2470]: Sounds good. I'll provide an updated pr when I am done with the new tests.

File: wallet/wallet.go [hunk 10]
```diff
@@ -2464,6 +2467,56 @@ func (w *Wallet) GetTransactions(startBlock, endBlock *BlockIdentifier,
...
+		txmgrNs := dbtx.ReadBucket(wtxmgrNamespaceKey)
+
+		txDetail, err := w.TxStore.TxDetails(txmgrNs, &txHash)
+		if err != nil {
+			return err
+		}
+
+		// If the transaction was not found we return an error.
+		if txDetail == nil {
+			return ErrNoTx
```
Comment (yyforyongyu) [inline wallet/wallet.go:2496]: nit: `return fmt.Errorf("%w: txid %v", ErrNoTx, txHash)`

File: wallet/wallet_test.go [hunk 11]
```diff
@@ -205,3 +206,89 @@ func TestLabelTransaction(t *testing.T) {
...
+
+	tests := []struct {
+		name string
+
+		// Whether the transaction should be known to the wallet.
+		txKnown bool
+
+		// Whether the transaction is mined.
+		minedTx bool
```
Comment (yyforyongyu) [inline wallet/wallet_test.go:221]: this field can be removed since it's not used.

File: wallet/wallet_test.go [hunk 12]
```diff
@@ -205,3 +206,89 @@ func TestLabelTransaction(t *testing.T) {
...
+						)
+
+						return w.TxStore.InsertTx(ns, rec, test.blockMeta)
+
+					})
+				require.NoError(t, err, "could not insert tx")
+			}
+
+			tx, err := w.GetTransaction(*TstTxHash)
+			if err != test.expectedErr {
```
Comment (yyforyongyu) [inline wallet/wallet_test.go:282]: can use `require.ErrorIs`

File: wallet/wallet_test.go [hunk 13]
```diff
@@ -205,3 +206,89 @@ func TestLabelTransaction(t *testing.T) {
...
+				t.Fatalf("expected: %v, got: %v", test.expectedErr, err)
+			}
+			if test.txKnown {
+				require.True(t, tx.Summary.Hash.IsEqual(TstTxHash))
+			}
+			// Test if the transaction is mined.
+			if test.blockMeta != nil {
```
Comment (yyforyongyu) [inline wallet/wallet_test.go:289]: This is testing that when it's mined, the `Height` field is set, but it's not testing when it's not mined, this field is not set.

File: wallet/wallet_test.go [hunk 14]
```diff
@@ -205,3 +206,89 @@ func TestLabelTransaction(t *testing.T) {
...
+	for _, test := range tests {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			w, cleanup := testWallet(t)
+			defer cleanup()
+
+			// If the transaction should be known to the store, we
+			// write txdetail to disk.
+			if test.txKnown {
```
Comment (yyforyongyu) [inline wallet/wallet_test.go:261]: this approach works, tho imo a better version would be creating testing a `txMined` and `txUnmined`, then define the table tests as 
```go
{
  name string
  txid chainhash.Hash
  expectedTx *GetTransactionResult
  expectedErr error
}
```
Then you construct three txids - `txMined = ...`, `txUnmined= ...`, and `txUnknown = ...`. And finally check return returned values from `GetTransaction`,
```go
require.ErrorIs(t, tc.expectedErr, err)
require.Equal(t, tc.expectedTx, tx)
```

Comment (ErikEk) [reply wallet/wallet_test.go]: @yyforyongyu I have updated the test code. I am not sure if you wanted me to add all transactions to the wallet storage  at once or one for each test cycle like I have done now?

Comment (ErikEk) [reply wallet/wallet_test.go]: I had issues storing the full GetTransactionResults array and opted for comparing the txid and block height as validators.

File: wallet/wallet.go [hunk 15]
```diff
@@ -2464,6 +2467,56 @@ func (w *Wallet) GetTransactions(startBlock, endBlock *BlockIdentifier,
...
+			res.Height = txDetail.Block.Height
+			res.BlockHash = &txDetail.Block.Hash
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &res, err
```
Comment (guggero) [inline wallet/wallet.go:2517]: nit: Here it should explicitly be `return &res, nil`.

**Review by guggero — APPROVED** [review]
Nice, LGTM :tada:

File: wallet/wallet.go (hunk 15 above)
Comment (ErikEk) [reply wallet/wallet.go]: Fixed.

File: wallet/wallet_test.go [hunk 16]
```diff
@@ -205,3 +216,90 @@ func TestLabelTransaction(t *testing.T) {
 		})
 	}
 }
+
+// TestGetTransaction tests if we can fetch a mined, an existing
+// and a non-existing transaction from the wallet like we expect.
+func TestGetTransaction(t *testing.T) {
+	t.Parallel()
+	rec, err := wtxmgr.NewTxRecord(TstSerializedTx, time.Now())
+	require.NoError(t, err)
```
Comment (yyforyongyu) [inline wallet/wallet_test.go:225]: nit: add a newline here

File: wallet/wallet_test.go [hunk 17]
```diff
@@ -205,3 +216,90 @@ func TestLabelTransaction(t *testing.T) {
...
+			},
+			expectedErr: ErrNoTx,
+		},
+	}
+	for _, test := range tests {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			w, cleanup := testWallet(t)
+			defer cleanup()
```
Comment (yyforyongyu) [inline wallet/wallet_test.go:283]: nit: add a newline here

**Review by yyforyongyu — APPROVED** [review]
LGTM🙏

**Comment by ErikEk** [comment]
@yyforyongyu PR updated.

**Pull request merged**